DietPi-Globals | Processing animation + bash.bashrc|occ|ncc into dietpi-globals

MichaIng opened the pull request:
+ Move bash.bashrc adjustments into dietpi-globals.
+ Create animation for processing messages.
+ Move ownCloud/Nextcloud occ/ncc into dietpi-globals as global functions.
+ Create optional software aliases/functions only, if target file exists.
+ Run occ/ncc via G_RUN_CMD

Commit:
v6.0

+ Move bash.bashrc adjustments into dietpi-globals.
+ Create animation for processing messages.

Commit:
v6.0

+ Remove bash.bashrc adjustments from set_core_environment

Commit:
v6.0

+ DietPi-Software | ownCloud/Nextcloud: Switch to occ/ncc alias

Commit:
v6.0

+ Start ownCloud/Nextcloud maintenance:mode with G_RUN_CMD

Commit:
v6.0

+ Fix missing line ending

Fourdee commented on dietpi/dietpi-software:
```diff
@@ -8664,8 +8663,7 @@ _EOF_
 			# Add occ command shortcut, use 'ncc' as 'occ' is reserved for ownCloud:
-			echo -e '#!/bin/bash\nsudo -u www-data php /var/www/nextcloud/occ $*' > /usr/local/bin/ncc
-			chmod +x /usr/local/bin/ncc
+			sed -i 's/#alias ncc/alias ncc/' /DietPi/dietpi/func/dietpi-globals
```
```dietpi-globals``` will be replaced during any ```dietpi-updates```. Any of these changes will be lost.

Only thing I can think of is either:
- Always have the alias in place (as long as the command returns 'not found' or similar, should be fine)
- Create a include file ```/var/lib/dietpi/dietpi-globals/alias_include``` which gets loaded with ```dietpi-globals```. If we plan to have vast number of software specific alias in future, this may be a better solution.
-- The include file could simply ```if [ -f /var/www/owncloud/occ ]; then``` allow alias? As long as it does not slow down script load.

MichaIng replied:
@Fourdee 
That's true, didn't think about that. I also saw there are already two static non core alias' inside: https://github.com/Fourdee/DietPi/pull/1377/files#diff-03a4faa1a3aeff0af60938482f5791b3R57

`if [ -f /var/www/owncloud/occ ]; then` could be by the way also used within main dietpi-globals (as well as for all other additional software), then there is one file less to check on script loading ;). If we choose a separate file, then I would use commenting, just because we can there, otherwise use dietpi-globals. One other thing is, that globals would need to be reloaded if the command wants to be used within installation/configuration (like in ownCloud/Nextcloud), on the other hand, just use `alias occ="..."` as command there also works until next globals reload. That way er keep it as installation step and can inform users 😉.

I will make some tests about script loading influence with: `time for i in ``seq 1 100``; do ; done`
€: Done, 100 executions in 1.70 - 1.75 seconds, no real difference if occ/ncc file check + alias creating is inside or commented out. I will just leave it inside global with file check for now and add a single alias command to installation script.

Fourdee commented:
@MichaIng 
very nice, only issue is with update = reset of ```dietpi-globals```

Commit:
v6.0

+ DietPi-Software | ownCloud/Nextcloud: Switch to file dependant occ alias

Commit:
v6.0

+ DietPi-Software | ownCloud/Nextcloud: Switch to file dependant occ alias

Commit:
v6.0

+ DietPi-Software | OpenTyrian: Switch to conditional alias

Commit:
v6.0

+ DietPi-Software | OpenTyrian: Switch to conditional alias

MichaIng commented:
Also made OpenTyrian a conditional alias. Didn't found software index for RetroPi emulation station, is this just present on related image?

Commit:
v6.0

+Switch Prefer_IPversion() to new file name.

MichaIng commented:
@Fourdee 
Just found:
`nano /etc/apt/apt.conf.d/99force-ipv`
> #empty line
>^[[KAcquire::ForceIPv4 "true";

Due to https://github.com/Fourdee/DietPi/blob/master/dietpi/func/dietpi-set_hardware#L1285-L1295, but I guess the additional garbage is coming from unset locale? Other topic...

For now I switched to new file name from preparation script. Do you think, forcing IPv6 is ever a good idea there? It literally sounds like "forcing" instead of "preferring". The latter would make sense to set, id one activates/proves IPv6 working, but as we know this is already default, if not IPv4 is forced, so not necessary to adjust. If it is really a "force" then this could break all APT connections to mirrors without IPv6 offer. Not sure, how many still exist, but I would not risk it.

k-plan commented:
@MichaIng 

> off topic

> It literally sounds like "forcing" instead of "preferring".

😄 

https://github.com/Fourdee/DietPi/issues/472

MichaIng commented:
@Fourdee 
Just reopened the issue. We need to find a more consistent way there, also switching the force-settings on toggle IPv6 within dietpi-config etc.

I found the issue with corrupted config above: My echo/printf hooks of course also affect echo to file. This is dangerous!! I will remove them and add killing the process animation to our other notification types.

k-plan commented:
@MichaIng 

sorry, will vote for disable your nice looking K.A.R.R Scanner feature at the moment.

![7u5jig](https://user-images.githubusercontent.com/17237601/34650755-7a02ef9a-f3c6-11e7-81ee-1ed398b0dde2.gif)

- test it on a RPi Zero W and  animated yellow dots will stay as last line on HDMI output sometimes

- some HDMI output messages `[   OK   ]` or not centered `[OK     ]` (loss this feature while booting up)

- some HDMI  output messages loss line break `[  .   ] [OK     ] Demo Message`
![180107-0004-c](https://user-images.githubusercontent.com/17237601/34650889-7de92c12-f3c8-11e7-8cb9-7e3bfc6c96ae.gif)

- some messages loss first character 
![180107-0007-c](https://user-images.githubusercontent.com/17237601/34650901-d04f8924-f3c8-11e7-8153-b13fa59f58c5.gif)

- and personally dislike, where is no `[   OK   ]` switching after successfull animated yellow dots processing
![180107-0006-c](https://user-images.githubusercontent.com/17237601/34650949-a2e4671a-f3c9-11e7-965c-e8b5c8373a8b.gif)

But just my  2 cents.
If we will work on it, this can be a very beneficial feature. 👍

Fourdee commented:
@MichaIng 

> Also made OpenTyrian a conditional alias. Didn't found software index for RetroPi emulation station, is this just present on related image?

Its to allow users who install retropie to run emulationstation under root user. The default install alias does not allow this.

Commit:
v6.0

+ DietPi-Globals | Processing: Remove echo/printf wrappers, as they have unexpected effects, e.g. adding garbage to echo > file.

Commit:
v6.0

+ tiny

MichaIng commented:
@k-plan 
Jep, known issues: https://github.com/Fourdee/DietPi/issues/1371#issuecomment-355758359
Would need further adjustments if otherwise wanted:
- As I added the animation also to dietpi commands headers (instead of the extra "Please wait..." line), we would need to take care, that first outputs of the scripts are also done via G_DIETPI-NOTIFY (if available there) or manually kill animation and overwrite line.
- As all APT output is done via (for me) unknown method, I don't know how to catch them to let stop animation/overwrite line.
  - Solution 1 (as mentioned within issue) would be to use e.g. G_RUN_CMD to run APT commands to suppress all terminal output (instead as current behaviour save to temp file and just print as log on error) until [Ok  ]/[Error] is used for result.
  - Solution 2 would be to not use processing messages (with animation) for APT commands, as they anyway have mostly constant output that shows what is going on/that something is going on.

About [Ok  ] not in the middle:
- As with Info, Ok, Error we have words with even and uneven numbers, there is no "middle" for all of them. My impression was, that it looked a bid strange, when [   Error   ] came not centred, while Ok/Info did. So I aligned all left, now that that brackets can stay closer together.
- Solution would be to find words, all with even or all uneven character number: Ok, Info and Failed e.g.?

About K.A.R.R Scanner: 😆 I could add a trailing darker dot to let it more look like the real K.I.T.T. , color fine tuning included? 🎉

Commit:
v6.0

+ DietPi-Software | ownCloud/Nextcloud: Revert to occ/ncc functions, as aliases are not accessible from non-interactive scripts.

Commit:
v6.0

DietPi-Software | ownCloud/Nextcloud: Revert to occ/ncc functions, as aliases are not accessible from non-interactive scripts.

k-plan commented:
@MichaIng 

> About [Ok ] not in the middle:

Ja, die Sprachbarriere ....

Es geht um die HDMI Ausgabe, im SSH Terminal kann ich damit sicher leben, da konsistent.

Der Wechsel sieht auf dem Monitor schon etwas merkwürdig aus, während des Bootvorgangs:

![2018-01-07 19-01-31](https://user-images.githubusercontent.com/17237601/34652468-c9afc1ae-f3de-11e7-9f19-56e65fe42a2b.gif)

Hoffe, man kann es erkennen.

edit;

das mit K.I.T.T. / K.A.R.R war ein Spass, hatte mich halt einfach daran erinnert, als ich es zum ersten mal gesehen habe. Nix für Ungut. 😄

MichaIng commented:
@k-plan 
Ah, das ist der Übergang zwischen Kernel/Boot und DietPi Meldungen, der hat ja aber noch nie gestimmt. Aber kann man natürlich bessern:

In English:
- The problem with boot/kernel -> DietPi messages is, that the kernel messages look different on every device, even seem to change between versions:
  - RPi was `[ ok ]/[FAIL]/[info]/[....]`, now it seems `[  OK  ]/[FAILED]/[ INFO ]`?
  - On my VM I don't see anything like this.
  - The systemd messages then look again totally different `[         <timestamp>] systemd:`, these are the the ones I get to see.
- The brackets start at left edge, instead 1 gap, that do have all in common.

We could adapt the RPi kernel way, for me it is also more beautiful, if the words are centred inside the brackets.

About K.I.T.T.:
- I'm all for things like this, bringing a little smile to peoples face. But I'm usually not creative enough, so great if I just hit one by chance 😄.
- And ah, just found K.A.R.R that's true, K.I.T.T.'s light was red.

Commit:
v6.0

+ Move aliases to dietpi-globalss

Commit:
v6.0

+ Move aliases to dietpi-globals.

Commit:
v6.0

tiny

Commit:
v6.0

tiny

MichaIng commented:
![k a r r](https://user-images.githubusercontent.com/28480705/34653879-679bd34e-f3f3-11e7-97b1-c22f8d8291e8.png)
Sadly there is no DietPi unicode symbol 😄.

But will stick to the dot, which looks more like "waiting".
- [x] Remove animation from G_AGX commands.
- [x] Remove animation on G_DIETPI-NOTIFY 3 due to random running animation after boot, even this should not be possible without related pid file.

Commit:
v6.0

+ DietPi-Globals | Process notification integration

Commit:
v6.0

+ DietPi-Globals | Process notification integration

Commit:
v6.0

+ DietPi-Globals | Process notification integration

Commit:
v6.0

+ DietPi-Globals | Process notification integration

Commit:
v6.0

+ fix

Commit:
v6.0

+ tiny

Commit:
v6.0

+ DietPi-Globals | Processing notification integration

Commit:
v6.0

+ DietPi-Globals | Revert Animation with G_DIETPI-NOTIFY 3

Commit:
v6.0

+ DietPi-Globals | Revert Animation with G_DIETPI-NOTIFY 3

Commit:
v6.0

DietPi-Globals | Revert animation on G_DIETPI-NOTIFY 3

Commit:
v6.0

DietPi-Globals | Fix processing animation

Commit:
v6.0

+ tiny

Commit:
v6.0

+ Fix quoting

MichaIng commented:
@k-plan @Fourdee 
Huii more work and trouble than I thought. If animation is started after dietpi-script header (G_DIETPI-NOTIFY 3) then somehow every ~2nd time an animation stays in place on boot, without any pid file in place to automatically kill it. This should be impossible, since every animation process is directly stored to pid file and while file is there, no additional process can be started. But yeah, maybe some special parallel magic, or different execution users during boot arranges that. So finally I removed animation from there. It was anyway a mess to handle this in every script+argument reasonable.

So for now animation is just started by G_RUN_CMD, root-, url- and file check, where we fully redirect every output and print an error handled result [  OK  ]/[FAILED].
All quirks, overwritten or displaced text should be gone. At least during my intense tests (software re/un/installation, image preparation) nothing like this showed up any more.

Stile follows now current RPi boot/kernel messages.

What I was stumbled about:
- Before this PR I was somehow able to globally access the Print_... functions inside dietpi-globals, even that they were created inside G_DIETPI-NOTIFY function.
- I used that to access the animation cleaning function from other scripts.
- After PR, where I also unset all these functions for that reasons in between, access to these function(s) was gone. Even after removing unsets again, reboot, new image preparation etc. no access to the functions any more, as I actually would also expect, as they are created within another function.

Using G_DIETPI-NOTIFY more consequently in our scripts, reducing/redirecting other outputs is now another task than can be done by times.

Fourdee commented:
@MichaIng 

> Before this PR I was somehow able to globally access the Print_... functions inside dietpi-globals, even that they were created inside G_DIETPI-NOTIFY function

Appears the Print functions are available after running ```G_DIETPI-NOTIFY```:
```
root@DietPi:~# G_DIETPI-NOTIFY 2 test
 [   Info   ] test
root@DietPi:~# Print_
Print_Failed        Print_Input_String  Print_Process
Print_Info          Print_Ok
```

We'll need to unset the Print_* functions in globals at end of notify. As they should only be available locally to ```G_DIETPI-NOTIFY```

EDIT: i see you have done that, but they are no commented?
https://github.com/Fourdee/DietPi/pull/1377/files#diff-03a4faa1a3aeff0af60938482f5791b3R290

> After PR, where I also unset all these functions for that reasons in between, access to these function(s) was gone. Even after removing unsets again, reboot, new image preparation etc. no access to the functions any more, as I actually would also expect, as they are created within another function.

I see the dilemma. Let me merge for testing.

Fourdee commented:
@MichaIng 

Legend 👍 

Merged for testing/further dev.

Fourdee commented:
@MichaIng 

Love the animation, brilliant work 👍

Fourdee commented:
Reference @MichaIng 
https://github.com/Fourdee/DietPi/commit/0edf8a2623de9ba0d269e4c558c6b764c3583329
